src/video/opengl.cpp

```diff
@@ -0,0 +1,1267 @@
+/* $Id$ */
+
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file opengl_v.cpp OpenGL video driver support. */
+
+#include "../stdafx.h"
+
+#ifdef WITH_OPENGL
+
+/* Define to disable buffer syncing. Will increase max fast forward FPS but produces artifacts. Mainly useful for performance testing. */
+// #define NO_GL_BUFFER_SYNC
+
+#if defined(_WIN32)
+#	include <windows.h>
+#else
+#	include <dlfcn.h>
+#endif
+
+#if defined(__APPLE__)
+#	include <OpenGL/gl3.h>
+#else
+#	include <GL/gl.h>
+#	if defined(UNIX)
+#		include <EGL/egl.h>
+#	endif
+#endif
+#include "../3rdparty/opengl/glext.h"
+
+#include "opengl.h"
+#include "../core/mem_func.hpp"
+#include "../core/math_func.hpp"
+#include "../core/mem_func.hpp"
+#include "../gfx_func.h"
+#include "../debug.h"
+#include "../blitter/factory.hpp"
+#include "../zoom_func.h"
+
+#include "../table/opengl_shader.h"
+
+
+#include "../safeguards.h"
+
+
+static PFNGLDEBUGMESSAGECONTROLPROC _glDebugMessageControl;
```

Comment:
Using autogenerated OpenGL loader such as https://github.com/Dav1dde/glad might be simpler that doing all that manually in our code.

Reply:
Yes. And most people will say to just chuck in GLEW.

Doing this for OTTD was only my second motivation, primarily I wanted something to get more familiar with OpenGL. And for that goal, I chose the do-it-yourself way, also called the know-whats-happening-behind-the-curtain-way.

It's certainly not set in stone, and if everybody else likes some kind of external dependency more, changing  it is possible.

Reply:
GLEW doesn't support OpenGL ES though (which we presumably will want to add because of WebGL/Android).